source/slides/11_chef.rst

```diff
@@ -0,0 +1,606 @@
+      block.call
+    end
+
+    foo do
+      puts 'hello!'
+    end # prints 'hello!'
+
+The ``foo(&block)`` declaration tells ruby that this argument takes a block which will be passed in later, and to convert that block into a proc
+
+Fake Chef
+---------
+
+You will notice chef syntax looks a lot like the last slide.
+
+.. code-block:: ruby
+
+    package 'vim' do
+      action :upgrade
+    end
+
+.. code-block:: ruby
+
+    def action(ac)
+      proc { |n| puts "apt-get #{ac} {n}"}
+    end
+    def package(n,&b) # n is just a regular old string
+      b.call.curry[n]
+    end
+    package 'vim' do
+      action :upgrade
+    end  # prints 'apt-get ugprade vim'
+
+In Chef ``action`` and other options are actually just symbols that get processed later.
+
+One Last Thing
+--------------
+
+``do end`` and ``{}`` are equivalent. Use ``do end`` for multiline blocks, and ``{}`` for single lines:
+
+.. code-block:: ruby
+
+    [1,2,3].inject(0) { |s,i| s += i }
+    {1:2, 3:4}.map do |k,v|
+      puts "k+v is #{k+v}"
+      puts "k*v is #{k*v}"
+    end
+
+Chef Components
+---------------
+
+* Cookbooks
+* Nodes
+* Roles
+* Environments
+* Data Bags
+
+Cookbooks
+---------
+
+The major components are:
+
+* Attributes
+* Recipes
+* Files/Templates
+* Libraries/Definitions (helpers, we won't cover this)
+* Lightweight Resource-Providers (we won't cover this)
+
+Attributes
+----------
+
+Can be defined in any of the following:
+
+* Cookbook
+* Node
+* Role
+* Environment
+
+There are 4 levels of attributes:
+
+* Default
+* Normal
+* Override
+* Automatic (special)
+
+Attributes (Cookbook)
+---------------------
+
+* Found in the ``attributes/`` dir in the root of a cookbook.
+
+.. code-block:: ruby
+
+    default['my_cookbook']['package_i_want'] = 'vim'
+
+* Often defined in recipes as well:
+
+.. code-block:: ruby
+
+    node.default['my_cookbook']['package_i_want'] = 'vim'
+
+Attributes can be accessed in a recipe like the following
+
+.. code-block:: ruby
+
+    node['my_cookbook']['package_i_want']
+
+Resources (Cookbook)
+---------------------
+
+* These are the workhorses of chef
+* Most things that you can do are defined via resources. Chef has a syntax for resources
+
+.. code-block:: ruby
+
+    resource 'name' do
+      option 'option_value'
+    end
+
+* Common resources used include: package, service, file, template,
+* Universal options include ``action, subscribes, notifies, only_if, not_if``
+
+Resource Examples
+-----------------
+
+.. code-block:: ruby
+
+    package 'apache2' do
+      action :install
+    end
+
+    package 'apache2' # the default action is :install
+
+    service 'apache2' do
+      action [:start, :enable]
+    end
+
+    template '/etc/apache2/sites-available/mysite.conf' do
+      source 'mysite.conf.erb'
+      owner 'wwwdata'
+      group 'wwwdata'
+      mode 0644 # like chmod (the 0 means octal in ruby)
+      notifies :restart, 'service[apache2]'
+      variables :some_other_var => 'example'
```

Comment:
Do we want to switch this to ruby 1.9 syntax? I know rubocop complains about this.